datadog/dogstatsd/base.py

```diff
@@ -274,24 +297,66 @@ def __init__(
         self._reset_telemetry()
         self._telemetry_flush_interval = telemetry_min_flush_interval
         self._telemetry = not disable_telemetry
+        self._last_flush_time = time.time()
 
+        self._current_buffer_total_size = 0
+        self._buffer = []  # type: List[Text]
         self._buffer_lock = RLock()
 
+        self._reset_buffer()
+
+        # This lock is used for backwards compatibility to prevent concurrent
+        # changes to the buffer when the user is managing the buffer themselves
+        # via deprecated `open_buffer()` and `close_buffer()` functions.
+        self._manual_buffer_lock = RLock()
+
+        # If buffering is disabled, we bypass the buffer function.
+        self._send = self._send_to_buffer
+        if disable_buffering is True:
+            log.info("Statsd buffering is disabled")
+            self._send = self._send_to_server
+
+        # Start the flush thread if buffering is enabled and the interval is above
+        # a reasonable range. This both prevents thrashing and allow us to use "0.0"
+        # as a value for disabling the automatic flush timer as well.
+        if disable_buffering is False and flush_interval >= MIN_FLUSH_INTERVAL:
+            self._register_flush_thread(flush_interval)
```

Comment:
If you disable the automatic flush thread, the assumption is that the user will be doing their own lifecycle management via `statsd_instance.flush()`

Reply:
> If you disable the automatic flush thread, the assumption is that the user will be doing their own lifecycle management via statsd_instance.flush()

What do you think about documenting this behavior (If not already done) as `flush_interval=0` could also means "flush automatically for every metric"?

> Also, the packet size splitter logic will send the data out when the packet is full regardless of if you have a background thread or not.

Consider the following code:
```python
options = {
    'statsd_host':'127.0.0.1',
    'statsd_port':8125
}

initialize(**options)
statsd.increment('my_application.start', tags=["environment:dev"])
long_running_method_that_does_not_use_statsd()
statsd.flush()
```
In the previous code, the metric `'my_application.start'` can be sent few minutes after the start of the application and let the user believes that the metric is not sent. This kind of code is common when testing the first time statsd (I had a couple of support cards related to that).

Reply:
> What do you think about documenting this behavior (If not already done) as flush_interval=0 could also means "flush automatically for every metric"?

Indeed - the documentation will be following this PR (AC-1040)

> In the previous code, the metric 'my_application.start' can be sent few minutes after the start of the application and let the user believes that the metric is not sent. This kind of code is common when testing the first time statsd (I had a couple of support cards related to that).

If I understand your comment I don't think this will be a problem here as `flush` is a new addition to the API and the delay would only be relevant if they set `disable_buffering=True` since the flush thread will send whatever partial-packet data we have automatically within the default interval (300ms) .